source/s3_checksums.c

```diff
@@ -0,0 +1,224 @@
+#include "aws/s3/private/s3_checksums.h"
+#include <aws/checksums/crc.h>
+
+size_t digest_size_from_algorithm(enum aws_s3_checksum_algorithm algorithm) {
+    switch (algorithm) {
+        case AWS_CRC32C:
+            return AWS_CRC32C_LEN;
+        case AWS_CRC32:
+            return AWS_CRC32_LEN;
+        case AWS_SHA1:
+            return AWS_SHA1_LEN;
+        case AWS_SHA256:
+            return AWS_SHA256_LEN;
+        case AWS_MD5:
+            return AWS_MD5_LEN;
+        default:
+            AWS_ASSERT(false);
+            return 0;
+    }
+}
+
+void s3_hash_destroy(struct aws_checksum *checksum) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    aws_hash_destroy(hash);
+    aws_mem_release(checksum->allocator, checksum->vtable);
+    aws_mem_release(checksum->allocator, checksum);
+}
+
+int s3_hash_update(struct aws_checksum *checksum, const struct aws_byte_cursor *to_checksum) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    return aws_hash_update(hash, to_checksum);
+}
+
+int s3_hash_finalize(struct aws_checksum *checksum, struct aws_byte_buf *output, size_t truncate_to) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    checksum->good = false;
+    return aws_hash_finalize(hash, output, truncate_to);
+}
+
+struct aws_checksum *aws_hash_new(struct aws_allocator *allocator, aws_hash_new_fn hash_fn) {
+    struct aws_checksum *checksum = aws_mem_acquire(allocator, sizeof(struct aws_checksum));
+    struct aws_checksum_vtable *vtable = aws_mem_acquire(allocator, sizeof(struct aws_checksum_vtable));
```

Comment:
It's not a huge deal, but allocating a vtable on every object creation is not something we tend to do and isn't really necessary given that everything is single inheritance.